src/lsm/tree.zig

```diff
@@ -564,70 +571,62 @@ pub fn TreeType(comptime TreeTable: type, comptime Storage: type, comptime tree_
                     tree.compact_mutable_table_into_immutable();
                 }
 
-                tree.compaction_callback = callback;
-                tree.grid.on_next_tick(compact_ready_next_tick, &tree.compaction_next_tick);
+                tree.compaction_callback = .{ .next_tick = callback };
+                tree.grid.on_next_tick(compact_finish_next_tick, &tree.compaction_next_tick);
                 return;
             }
             assert(op == tree.lookup_snapshot_max);
 
-            tree.compact_start(callback);
-            tree.compact_drive();
-        }
-
-        fn compact_ready_next_tick(next_tick: *Grid.NextTick) void {
-            const tree = @fieldParentPtr(Tree, "compaction_next_tick", next_tick);
-            assert(tree.compaction_io_pending == 0);
-
-            const callback = tree.compaction_callback.?;
-            tree.compaction_callback = null;
-            callback(tree);
-        }
-
-        fn compact_start(tree: *Tree, callback: fn (*Tree) void) void {
-            assert(tree.compaction_io_pending == 0);
-            assert(tree.compaction_callback == null);
-
-            if (constants.verify) {
-                tree.manifest.verify(tree.compaction_op);
-            }
-
-            tracer.start(
-                &tree.tracer_slot,
-                .{ .tree_compaction_beat = .{ .tree_name = tree_name } },
-                @src(),
-            );
-
-            tree.compaction_callback = callback;
-
             const compaction_beat = tree.compaction_op % constants.lsm_batch_multiple;
             const start = (compaction_beat == 0) or
                 (compaction_beat == half_bar_beat_count);
+            const end = (compaction_beat == half_bar_beat_count - 1) or
+                (compaction_beat == constants.lsm_batch_multiple - 1);
+            assert(!(start and end));
 
             const op_min = compaction_op_min(tree.compaction_op);
             assert(op_min < snapshot_latest);
             assert(op_min % half_bar_beat_count == 0);
 
-            log.debug(tree_name ++ ": compact_start: op={d} op_min={d} beat={d}/{d}", .{
+            log.debug(tree_name ++ ": compact: op={d} op_min={d} beat={d}/{d}", .{
                 tree.compaction_op,
                 op_min,
                 compaction_beat + 1,
                 constants.lsm_batch_multiple,
             });
 
-            if (start) tree.manifest.reserve();
+            if (start) {
+                if (constants.verify) {
+                    tree.manifest.verify(tree.compaction_op);
+                }
+
+                tree.manifest.reserve();
```

Comment:
Might be easier in the next one. I'm not sure how easy it will be to calculate the number of compactions up front though. Do all the reservations have to happen at the start of the half-bar? Is anything else writing to the grid outside of compaction?

Reply:
> Do all the reservations have to happen at the start of the half-bar?

The key here is that if a bunch of different actions need reservations (e.g. multiple concurrent compactions across trees/levels), then those reservations must be invoked in a deterministic order. (Otherwise the storage layout becomes non-deterministic).

> Is anything else writing to the grid outside of compaction?

There's grid repair, but that won't need reservations, since it only writes to already-reserved blocks.
There's state transfer, but that won't run concurrently with compaction.